### PR TITLE
[Part II] Implement ES2015 module system 

### PR DESCRIPTION
**Pull request by vincedani**
This is the 2nd part of the ES2015 module system implementation. Depends on the [Part I](https://github.com/pando-project/jerryscript/pull/2599).

Additional features to the Part I:
 - Implement whole module imports.
 - Support classes.
 - Import scripts for their side effects only.
 - Default imports and exports.

### CODE REVIEW

File: jerry-core/ecma/base/ecma-module.c [hunk 1]
```diff
@@ -13,246 +13,864 @@
...
-  {
-    parser_module_names_t *next_p = current_p->next_p;
+/**
+ *  Insert a {module, export_name} record into a resolve set.
+ *  Note: See 15.2.1.16.3 - resolveSet and exportStarSet
+ *
+ *  @return true - if the set already contains the record
+ *          false - otherwise
+ */
+bool ecma_module_resolve_set_insert (ecma_module_resolve_set_t **set_p, /**< resolve set */
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: `/**< [out] resolve set */`

Comment (LaszloLango) [reply jerry-core/ecma/base/ecma-module.c]: missing line break after `bool`

File: jerry-core/ecma/base/ecma-module.c [hunk 2]
```diff
@@ -13,246 +13,864 @@
...
+                                     ecma_module_t *module_p, /**< module */
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: It looks to me `module_p` and  `export_name_p` could be constant pointers.

File: jerry-core/ecma/base/ecma-module.c [hunk 3]
```diff
@@ -13,246 +13,864 @@
...
 
 /**
- * Connect the imported script's properties to the main script.
+ * Clean up contents of a resolve set.
  */
-static void
-ecma_module_connect_properties (ecma_object_t *scope_p) /** scope_p */
+void ecma_module_resolve_set_cleanup (ecma_module_resolve_set_t *set_p) /**< resolve set */
+{
+  ecma_module_resolve_set_t *current_p = set_p;
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: This new variable looks useless to me.

File: jerry-core/ecma/base/ecma-module.c [hunk 4]
```diff
@@ -13,246 +13,864 @@
...
+    jmem_heap_free_block (current_p, sizeof (ecma_module_resolve_set_t));
+    current_p = next_p;
+  }
+} /* ecma_module_resolve_set_cleanup */
+
+/**
+ * Pushes a new resolve frame on top of a resolve stack and initializes it
+ * to begin resolving the specified exported name in the base module.
+ */
+void ecma_module_resolve_stack_push (ecma_module_resolve_stack_t **stack_p, /**< resolve stack */
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: missing `[in,out]` from comment and line break after return value:
```c
void
ecma_module_resolve_stack_push (ecma_module_resolve_stack_t **stack_p, /**< [in,out] resolve stack */
                                ecma_module_t *module_p, /**< base module */
                                ecma_string_t *export_name_p) /**< exported name */
```

File: jerry-core/ecma/base/ecma-module.c [hunk 5]
```diff
@@ -13,246 +13,864 @@
...
-  parser_module_context_t *module_context_p = JERRY_CONTEXT (module_top_context_p);
+  new_frame_p->next_p = *stack_p;
+  *stack_p = new_frame_p;
+} /* ecma_module_resolve_stack_push */
 
-  if (module_context_p == NULL || module_context_p->exports_p == NULL || module_context_p->imports_p == NULL)
+/**
+ * Pop the topmost frame from a resolve stack.
+ */
+void ecma_module_resolve_stack_pop (ecma_module_resolve_stack_t **stack_p) /**< resolve stack */
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: ditto, missing '`[in,out]`' from comment and line break after return value

File: jerry-core/ecma/base/ecma-module.c [hunk 6]
```diff
@@ -13,246 +13,864 @@
...
+ * Resolves which module satisfies an export based from a specific module in the import tree.
+ * If no error occurs, out_record_p will contain a {module, local_name} record, which satisfies
+ * the export, or {NULL, NULL} if the export is ambiguous.
+ * Note: See 15.2.1.16.3
+ *
+ * @return ECMA_VALUE_ERROR - if an error occured
+ *         ECMA_VALUE_EMPTY - otherwise
+ */
+static ecma_value_t
+ecma_module_resolve_export (ecma_module_t *module_p, /**< base module */
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: Can `module_p` be a constant pointer here?

File: jerry-core/ecma/base/ecma-module.c [hunk 7]
```diff
@@ -13,246 +13,864 @@
...
+
+        indirect_export_p = indirect_export_p->next_p;
+      }
 
-        ecma_named_data_property_assign_value (global_obj_p,
-                                               ECMA_PROPERTY_VALUE_PTR (new_property_p),
-                                               prop_pair_p->values[i].value);
+      /* We need to check whether the indirect exports resolve to anything. */
+      continue;
+    }
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: Is this indentation correct? Would be better to mark with a comment which loop ends with this right brace.

File: jerry-core/ecma/base/ecma-module.c [hunk 8]
```diff
@@ -13,246 +13,864 @@
...
+                                         ECMA_PROPERTY_VALUE_PTR (new_property_p),
+                                         prop_value);
 
+  ecma_free_value (prop_value);
+  return result;
+} /* ecma_module_namespace_object_add_export_if_needed */
 
-  if (jerry_value_is_error (ret_value))
+/**
+ * Create a namespace object for a module.
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: `Creates...`

File: jerry-core/ecma/base/ecma-module.c [hunk 9]
```diff
@@ -13,246 +13,864 @@
...
 
-  ecma_module_add_lex_env (scope_p);
-  parser_module_free_saved_names (module_context_p->exports_p);
-} /* ecma_module_connect_properties */
+  return ret_value;
+} /* ecma_module_resolve_export */
 
 /**
- * Run an EcmaScript module loaded by ecma_module_load_modules.
+ * Resolve an export and add to the modules namespace object, if the export name is not yet handled.
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: `Resolves...`

File: jerry-core/ecma/base/ecma-module.c [hunk 10]
```diff
@@ -13,246 +13,864 @@
...
+  }
 
-  JERRY_ASSERT (ecma_get_object_type (func_obj_p) == ECMA_OBJECT_TYPE_FUNCTION);
+  return result;
+} /* ecma_module_create_namespace_object */
 
-  ecma_object_t *scope_p = ecma_create_decl_lex_env (ecma_get_global_environment ());
-  ret_value = vm_run_module (ecma_op_function_get_compiled_code ((ecma_extended_object_t *) func_obj_p), scope_p);
+/**
+ * Evaluate an EcmaScript module.
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.c]: Evaluates

File: jerry-core/ecma/base/ecma-module.h [hunk 11]
```diff
@@ -16,12 +16,125 @@
...
+
+/**
+ * Module structure storing an instance of a module
+ */
+struct ecma_module
+{
+  struct ecma_module *next_p;         /**< next linked list node */
+  ecma_module_state_t state;          /**< state of the mode */
+  ecma_string_t *path_p;              /**< path of the module */
+  ecma_module_context_t *context_p;   /**< module context of the module */
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-module.h]: I do not like this kind of cycling reference. Why is this necessary?

File: jerry-core/parser/js/js-parser-module.c [hunk 12]
```diff
@@ -173,393 +199,275 @@ parser_module_add_export_node_to_context (parser_context_t *context_p) /**< pars
...
+        && lexer_compare_raw_identifier_to_current (context_p, "from", 4))
+    {
+      parser_raise_error (context_p, PARSER_ERR_RIGHT_BRACE_EXPECTED);
+    }
   }
-} /* parser_module_handle_requests */
+} /* parser_module_parse_import_clause */
 
 /**
  * Raises parser error if the import or export statement is not in the global scope.
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser-module.c]: The builtin `Function(args, src)` also can trigger the parser. Is this handled below?

File: jerry-core/ecma/base/ecma-module.h (hunk 11 above)
Comment (dbatyai) [reply jerry-core/ecma/base/ecma-module.h]: It's not necessary in the strict sense of the word, but it's very useful for not having to look up which module the current context belongs to when connecting properties.

File: jerry-core/ecma/base/ecma-module.c (hunk 7 above)
Comment (dbatyai) [reply jerry-core/ecma/base/ecma-module.c]: Looks right to me, I'll add a comment nevertheless.

File: targets/nuttx-stm32f4/jerry_main.c [hunk 13]
```diff
@@ -558,6 +558,26 @@ jerry_port_release_source (uint8_t *buffer_p) /**< buffer to free */
   free (buffer_p);
 } /* jerry_port_release_source */
 
+/**
+ * Normalize a file path
+ *
+ * @return length of the path written to the output buffer
+ */
+size_t
+jerry_port_normalize_path (const char *in_path_p, /**< input file path */
```
Comment (LaszloLango) [inline targets/nuttx-stm32f4/jerry_main.c]: Please add the same to `targets/tizenrt-artik053/apps/jerryscript/jerry_main.c`.

Comment (dbatyai) [reply targets/nuttx-stm32f4/jerry_main.c]: Done.
